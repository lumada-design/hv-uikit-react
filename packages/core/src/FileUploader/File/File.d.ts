import * as React from "react";
import { StandardProps } from "@material-ui/core";

export interface File {
  /**
   * The file name.
   */
  name: string;
  /**
   * The upload status.
   */
  status: "progress" | "success" | "fail";
  /**
   * The file size in bytes.
   */
  size?: number;
  /**
   * Upload progress in bytes.
   */
  progress?: number;
}

export type FileEvent = (file: File) => void;

export type HvFileClassKey =
  | "progressbar"
  | "progressbarBack"
  | "nameText"
  | "progressTextContainer"
  | "removeButton";

export interface FileProps
  extends StandardProps<React.HTMLAttributes<HTMLDivElement>, HvFileClassKey> {
  /**
   * File information to be displayed
   */
  data: File;
  /**
   * Callback fired when file is removed from list.
   */
  onFileRemoved?: FileEvent;
  /**
   * Value of aria-label to apply to remove file button in filelist
   * */
  removeFileButtonLabel: string;
}

export default function HvFile(props: FileProps): JSX.Element | null;
