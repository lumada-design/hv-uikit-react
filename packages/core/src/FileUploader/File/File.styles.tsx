import { createClasses } from "@hitachivantara/uikit-react-utils";
import { theme } from "@hitachivantara/uikit-styles";

export const { staticClasses, useClasses } = createClasses("HvFile", {
  root: {
    position: "relative",
    display: "flex",
    gap: theme.space.xs,
    alignItems: "center",
    backgroundColor: theme.colors.bgContainer,
    padding: theme.space.xs,
    border: `1px solid ${theme.colors.border}`,
    borderRadius: `0px 0px ${theme.radii.round} ${theme.radii.round}`,
  },
  progressbar: {
    position: "absolute",
    top: "-1px",
  },
  progressbarContainer: {
    height: "4px",
  },
  progressbarBack: {},
  nameText: {
    display: "flex",
    alignItems: "center",
    flex: 1,
  },
  progressTextContainer: {
    display: "flex",
    flexGrow: 1,
    alignItems: "center",
  },
  removeButton: {},
  previewContainer: {
    display: "flex",
    width: "48px",
    height: "48px",
    justifyContent: "center",
    alignItems: "center",

    "& span": {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      width: "100%",
      height: "100%",
    },

    "& img": {
      width: "40px",
      height: "40px",
      objectFit: "cover",
      objectPosition: "center",
      alignSelf: "center",
    },
  },
  icon: {
    margin: theme.space.xs,
  },
  fail: {},
});
