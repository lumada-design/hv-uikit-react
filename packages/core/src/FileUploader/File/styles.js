const styles = (theme) => ({
  root: {},
  progressbarBack: {
    position: "absolute",
    top: "-1px",
    width: "100%",
    border: `1px solid ${theme.hv.palette.atmosphere.atmo4}`,
  },
  progressbar: {
    position: "absolute",
    top: "-1px",
    width: "80%",
    height: 2,
    border: `1px solid ${theme.hv.palette.accent.acce1}`,
    "&::-moz-progress-bar": {
      background: theme.hv.palette.accent.acce1,
    },
  },
  progressTextContainer: {
    display: "flex",
    flexGrow: 1,
    alignItems: "center",
  },
  previewContainer: {
    display: "flex",
    margin: theme.hvSpacing(0, "xs"),
    width: 52,
    height: 52,
    justifyContent: "center",
    alignItems: "center",

    "& img": {
      width: "100%",
      height: "100%",
      objectFit: "cover",
      objectPosition: "center",
    },
  },
  icon: {
    width: 32,
    height: 32,
    margin: theme.hvSpacing(0, "xs"),
  },
  removeButton: {
    margin: theme.hvSpacing(0, "xs"),
  },
  fail: {
    color: theme.palette.sema4,
  },
  nameText: {},
});

export default styles;
