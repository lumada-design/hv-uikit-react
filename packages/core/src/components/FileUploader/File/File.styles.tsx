import { theme } from "@hitachivantara/uikit-styles";

import { createClasses } from "@core/utils/classes";

export const { staticClasses, useClasses } = createClasses("HvFile", {
  root: {},
  progressbar: {
    position: "absolute",
    top: "-1px",
  },
  progressbarContainer: {
    height: "4px",
  },
  progressbarBack: {},
  nameText: {
    overflow: "hidden",
    textOverflow: "ellipsis",
    whiteSpace: "nowrap",
  },
  progressTextContainer: {
    display: "flex",
    flexGrow: 1,
    alignItems: "center",
  },
  removeButton: {
    width: 32,
    height: 32,
    margin: `0px ${theme.space.xs}`,
  },
  previewContainer: {
    display: "flex",
    margin: `0px ${theme.space.xs}`,
    width: "48px",
    height: "48px",
    justifyContent: "center",
    alignItems: "center",

    "& span": {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      width: "100%",
      height: "100%",
    },

    "& img": {
      width: "40px",
      height: "40px",
      objectFit: "cover",
      objectPosition: "center",
      alignSelf: "center",
    },
  },
  icon: {
    width: 32,
    height: 32,
    margin: `0px ${theme.space.xs}`,
  },
  fail: {},
});
