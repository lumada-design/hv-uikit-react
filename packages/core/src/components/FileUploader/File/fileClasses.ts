import { getClasses } from "@core/utils";

export interface HvFileClasses {
  root?: string;
  progressbar?: string;
  progressbarBack?: string;
  nameText?: string;
  progressTextContainer?: string;
  removeButton?: string;
  previewContainer?: string;
  icon?: string;
  fail?: string;
}

const classKeys: (keyof HvFileClasses)[] = [
  "root",
  "progressbar",
  "progressbarBack",
  "nameText",
  "progressTextContainer",
  "removeButton",
  "previewContainer",
  "icon",
  "fail",
];

const fileClasses = getClasses(classKeys, "HvFile");

export default fileClasses;
