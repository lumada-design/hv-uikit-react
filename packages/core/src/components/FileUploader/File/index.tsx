export { default as fileClasses } from "./fileClasses";
export * from "./fileClasses";
export * from "./File";
