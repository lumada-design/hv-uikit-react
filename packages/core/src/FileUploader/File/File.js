import React from "react";
import PropTypes from "prop-types";
import { withStyles } from "@mui/styles";
import { IconButton } from "@mui/material";
import { Fail, Close, Success } from "@hitachivantara/uikit-react-icons";
import HvTypography from "../../Typography";
import { setId } from "../../utils/setId";
import { convertUnits } from "../utils";
import styles from "./styles";

const getStatusIcon = (classes, status) => {
  switch (status) {
    case "success":
      return <Success className={classes.icon} color="sema1" />;
    case "fail":
      return <Fail className={classes.icon} color="sema4" />;
    default:
      return <div className={classes.icon} />;
  }
};

const getProgressText = (classes, data) => {
  const hasFailed = data.status === "fail";
  const inProgress = data.status === "progress";

  return (
    <>
      {data.progress || data.size || data.errorMessage ? `\xa0|\xa0` : null}

      {inProgress && data.progress != null && (
        <HvTypography variant="highlightText">
          {`${convertUnits(data.progress)}\xa0/\xa0`}
        </HvTypography>
      )}

      {!hasFailed && data.size && <HvTypography>{`${convertUnits(data.size)}`}</HvTypography>}

      {hasFailed && data.errorMessage && (
        <HvTypography className={classes.fail}>{data.errorMessage}</HvTypography>
      )}
    </>
  );
};

const getProgressBarWith = ({ size, progress }) => {
  const width = Math.round((progress * 100) / size);

  return width;
};

const File = ({ id, classes, data, onFileRemoved, removeFileButtonLabel }) => {
  const hasError = data.status === "fail";
  const inProgress = data.status === "progress";
  const progressText = getProgressText(classes, data);
  const statusIcon = getStatusIcon(classes, data.status);

  const currentProgress = getProgressBarWith(data);

  return (
    <li className={classes.root}>
      {!hasError && inProgress && <span className={classes.progressbarBack} />}

      {!hasError && inProgress && (
        <progress
          className={classes.progressbar}
          aria-valuemin="0"
          aria-valuemax="100"
          aria-valuenow={currentProgress}
          style={{ width: `${currentProgress}%` }}
        />
      )}

      {statusIcon}

      <HvTypography noWrap className={classes.nameText} variant="highlightText">
        {data.name}
      </HvTypography>

      <span className={classes.progressTextContainer}>{progressText}</span>

      {data.preview && <div className={classes.previewContainer}>{data.preview}</div>}

      <IconButton
        id={setId(id, "remove-button")}
        aria-label={removeFileButtonLabel}
        className={classes.removeButton}
        category="ghost"
        onClick={() => onFileRemoved?.(data)}
      >
        <Close iconSize="XS" />
      </IconButton>
    </li>
  );
};

File.propTypes = {
  /**
   * Id to be applied to the root node.
   */
  id: PropTypes.string,
  /**
   * A Jss Object used to override or extend the styles applied to the Switch Component.
   */
  classes: PropTypes.shape({
    /**
     * Styles applied to the root element.
     */
    root: PropTypes.string,
    /**
     * Style applied to the progress bar.
     */
    progressbar: PropTypes.string,
    /**
     * Style applied to the progress bar background.
     */
    progressbarBack: PropTypes.string,
    /**
     * Style applied to the file name.
     */
    nameText: PropTypes.string,
    /**
     * Style applied to the container of the progress bar.
     */
    progressTextContainer: PropTypes.string,
    /**
     * Style applied to the remove button.
     */
    removeButton: PropTypes.string,
    /**
     * Style applied to the file preview container.
     */
    previewContainer: PropTypes.string,
  }).isRequired,
  /**
   * File information to be displayed
   */
  data: PropTypes.shape({
    /**
     * The file id.
     */
    id: PropTypes.string,
    /**
     * The file name.
     */
    name: PropTypes.string,
    /**
     * The upload status.
     */
    status: PropTypes.oneOf(["progress", "success", "fail"]),
    /**
     * Optional node representing a preview of the uploaded file.
     */
    preview: PropTypes.node,
  }).isRequired,
  /**
   * Callback fired when file is removed from list.
   */
  onFileRemoved: PropTypes.func.isRequired,
  /**
   * Value of aria-label to apply to remove file button in filelist
   * */
  removeFileButtonLabel: PropTypes.string.isRequired,
};

export default withStyles(styles, { name: "HvFileUploaderFile" })(File);
