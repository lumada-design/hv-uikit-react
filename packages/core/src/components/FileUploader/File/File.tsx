import { Close, Fail, Success } from "@hitachivantara/uikit-react-icons";

import { setId } from "@core/utils/setId";
import { HvButton } from "@core/components/Button";
import { HvTypography } from "@core/components/Typography";
import { ExtractNames } from "@core/utils/classes";
import { HvProgressBar } from "@core/components/ProgressBar";
import { useDefaultProps } from "@core/hooks/useDefaultProps";

import { convertUnits } from "../utils";
import { staticClasses, useClasses } from "./File.styles";

export { staticClasses as fileClasses };

export type HvFileClasses = ExtractNames<typeof useClasses>;

export interface HvFileData extends Omit<File, "name" | "size"> {
  /**
   * The file id.
   */
  id?: string;
  /**
   * The file name.
   */
  name?: string;
  /**
   * The upload status.
   */
  status?: "progress" | "success" | "fail";
  /**
   * The file size in bytes.
   */
  size?: number;
  /**
   * Upload progress in bytes.
   */
  progress?: number;
  /**
   * Optional node representing a preview of the uploaded file.
   */
  preview?: React.ReactNode;
  /**
   * Error message when the upload failed.
   */
  errorMessage?: string;
}

export type HvFilesAddedEvent = (files: HvFileData[]) => void;

export type HvFileRemovedEvent = (file: HvFileData) => void;

export type HvFileProps = {
  /**
   * Id to be applied to the root node.
   */
  id?: string;
  /**
   * File information to be displayed.
   */
  data: HvFileData;
  /**
   * Callback fired when file is removed from list.
   */
  onFileRemoved?: HvFileRemovedEvent;
  /**
   * Value of aria-label to apply to remove file button in file list.
   * */
  removeFileButtonLabel?: string;
  /**
   * A Jss Object used to override or extend the styles applied to the component.
   */
  classes?: HvFileClasses;
};

const getStatusIcon = (
  classes?: HvFileClasses,
  status?: HvFileData["status"]
) => {
  switch (status) {
    case "success":
      return <Success className={classes?.icon} color="positive" />;
    case "fail":
      return <Fail className={classes?.icon} color="negative" />;
    default:
      return <div className={classes?.icon} />;
  }
};

const getProgressText = (data: HvFileData, classes?: HvFileClasses) => {
  const hasFailed = data.status === "fail";
  const inProgress = data.status === "progress";

  return (
    <>
      {data.progress || data.size || data.errorMessage ? `\xa0|\xa0` : null}

      {inProgress && data.progress != null && (
        <HvTypography variant="label">
          {`${convertUnits(data.progress)}\xa0/\xa0`}
        </HvTypography>
      )}

      {!hasFailed && data.size && (
        <HvTypography>{`${convertUnits(data.size)}`}</HvTypography>
      )}

      {hasFailed && data.errorMessage && (
        <HvTypography className={classes?.fail}>
          {data.errorMessage}
        </HvTypography>
      )}
    </>
  );
};

const getProgressBarWith = ({ size, progress }: HvFileData) => {
  const width =
    progress != null && size != null ? Math.round((progress * 100) / size) : 0;

  return width;
};

export const HvFile = (props: HvFileProps) => {
  const {
    id,
    classes: classesProp,
    data,
    onFileRemoved,
    removeFileButtonLabel,
  } = useDefaultProps("HvFile", props);
  const { classes, cx } = useClasses(classesProp);

  const hasError = data.status === "fail";

  const inProgress = data.status === "progress";

  const progressText = getProgressText(data, classes);

  const statusIcon = getStatusIcon(classes, data.status);

  const currentProgress = getProgressBarWith(data);

  return (
    <li className={classes.root}>
      {!hasError && inProgress && (
        <HvProgressBar
          classes={{
            root: classes.progressbar,
            progressBarContainer: cx(
              classes.progressbarContainer,
              classes.progressbarBack
            ),
          }}
          value={currentProgress}
          hideLabel
        />
      )}

      {statusIcon}

      <HvTypography className={classes.nameText} variant="label">
        {data.name}
      </HvTypography>

      <span className={classes.progressTextContainer}>{progressText}</span>

      {data.preview && (
        <div className={classes.previewContainer}>{data.preview}</div>
      )}

      <HvButton
        id={setId(id, "remove-button")}
        aria-label={removeFileButtonLabel}
        className={classes.removeButton}
        onClick={() => onFileRemoved?.(data)}
        icon
      >
        <Close iconSize="XS" />
      </HvButton>
    </li>
  );
};
