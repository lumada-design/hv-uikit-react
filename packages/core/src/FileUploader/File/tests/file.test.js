/* eslint-env jest */

import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";

import { HvProvider } from "../../..";
import File from "..";

const dataFail = {
  name: "somefile.png",
  size: 875,
  errorMessage: "File type not allowed for upload",
  progress: 25,
  status: "fail",
  fileStatus: "fail",
};

const dataSuccess = {
  name: "somefile.png",
  size: 875,
  errorMessage: "File type not allowed for upload",
  progress: 25,
  status: "success",
  fileStatus: "success",
};

const setup = ({ ...fileProps }) =>
  render(
    <HvProvider cssBaseline="none">
      <File
        onFilesAdded={() => {}}
        onFileRemoved={() => {}}
        removeFileButtonLabel="removeFileButtonLabel"
        {...fileProps}
      />
    </HvProvider>
  );

describe("File invalid", () => {
  it("renders correctly", () => {
    setup({ data: dataFail });

    const file = screen.getByRole("listitem");
    expect(file).toBeVisible();
    expect(file).toMatchSnapshot();
  });
});

describe("File valid", () => {
  it("renders correctly", () => {
    setup({ data: dataSuccess, unit: "mb" });

    const file = screen.getByRole("listitem");
    expect(file).toBeVisible();
    expect(file).toMatchSnapshot();
  });

  it("calls the delete file callback", () => {
    const onClickCallback = jest.fn();
    setup({ data: dataSuccess, unit: "mb", onFileRemoved: onClickCallback });

    const fileButton = screen.getByRole("button");

    fireEvent.click(fileButton);
    expect(onClickCallback).toHaveBeenCalled();
  });
});
