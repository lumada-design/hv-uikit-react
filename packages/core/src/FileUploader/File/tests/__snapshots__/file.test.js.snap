// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File invalid renders correctly 1`] = `
<li
  class="HvFileUploaderFile-root"
>
  <div
    class="HvFileUploaderFile-icon HvIconBase-root HvIconBase-s"
    name="Fail"
  >
    <svg
      focusable="false"
      height="16"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class="color0"
        d="M7.5 4.001h1v6h-1zm0 8h1v-1h-1zM16 8a8 8 0 10-8 8 8 8 0 008-8zm-1 0a7 7 0 11-7-7 7 7 0 017 7z"
        fill="#C62828"
      />
    </svg>
  </div>
  <p
    class="HvFileUploaderFile-nameText HvTypography-root HvTypography-highlightText HvTypography-noWrap"
  >
    somefile.png
  </p>
  <span
    class="HvFileUploaderFile-progressTextContainer"
  >
     | 
    <p
      class="HvFileUploaderFile-fail HvTypography-root HvTypography-normalText"
    >
      File type not allowed for upload
    </p>
  </span>
  <button
    aria-label="removeFileButtonLabel"
    category="ghost"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium HvFileUploaderFile-removeButton css-5xhalq-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <div
      class="HvIconBase-root HvIconBase-xs"
      name="Close"
    >
      <svg
        focusable="false"
        height="12"
        viewBox="0 0 16 16"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="color0"
          d="M8.7 8l5.3 5.3-.7.7L8 8.7 2.7 14l-.7-.7L7.3 8 2 2.7l.7-.7L8 7.3 13.3 2l.7.7z"
          fill="#414141"
        />
      </svg>
    </div>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</li>
`;

exports[`File valid renders correctly 1`] = `
<li
  class="HvFileUploaderFile-root"
>
  <div
    class="HvFileUploaderFile-icon HvIconBase-root HvIconBase-s"
    name="Success"
  >
    <svg
      focusable="false"
      height="16"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class="color0"
        d="M6.387 12L4 9.613l.707-.707 1.613 1.613 4.65-5.58.768.641zM8 15a6.953 6.953 0 01-4.943-2.057A7.023 7.023 0 118 15m8-7a8 8 0 10-8 8 8 8 0 008-8z"
        fill="#59941B"
      />
    </svg>
  </div>
  <p
    class="HvFileUploaderFile-nameText HvTypography-root HvTypography-highlightText HvTypography-noWrap"
  >
    somefile.png
  </p>
  <span
    class="HvFileUploaderFile-progressTextContainer"
  >
     | 
    <p
      class="HvTypography-root HvTypography-normalText"
    >
      875.00B
    </p>
  </span>
  <button
    aria-label="removeFileButtonLabel"
    category="ghost"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium HvFileUploaderFile-removeButton css-5xhalq-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <div
      class="HvIconBase-root HvIconBase-xs"
      name="Close"
    >
      <svg
        focusable="false"
        height="12"
        viewBox="0 0 16 16"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="color0"
          d="M8.7 8l5.3 5.3-.7.7L8 8.7 2.7 14l-.7-.7L7.3 8 2 2.7l.7-.7L8 7.3 13.3 2l.7.7z"
          fill="#414141"
        />
      </svg>
    </div>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</li>
`;
